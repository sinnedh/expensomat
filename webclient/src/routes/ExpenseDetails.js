import React from 'react'
import { connect } from 'react-redux'
import { EditableInput } from '../components/EditableField'
import { deleteExpense, updateExpense } from '../actions'

class ExpenseDetails extends React.Component {
  render() {
    return (
      <div>
        <h1>Expense details:</h1>
        <div>
          <label>Description:</label>{' '}
          <EditableInput
            isEditable={this.props.canEditExpenses}
            value={this.props.description}
            onClickSave={value => this.props.onUpdateDescription(value)}
            />
        </div>
        <div>
          <label>Amount:</label>{' '}
          <EditableInput
            isEditable={this.props.canEditExpenses}
            value={this.props.amount}
            onClickSave={value => this.props.onUpdateAmount(value)}
            />
        </div>
        <div>
          <label>Paid at:</label>{' '}
          <EditableInput
            isEditable={this.props.canEditExpenses}
            value={this.props.paid_at}
            onClickSave={value => this.props.onUpdatePaidAt(value)}
            />
        </div>
        {this.props.canEditExpenses && this.props.expenseId != null && 
          <button onClick={(e) => this.props.onClickDelete(e, this.props.expenseId)}>
            Delete
          </button>
        }
      </div>
    )
  }
}

const mapStateToProps = (state, ownProps) => {
  const expense = state.getIn(['expenses', ownProps.match.params.id.toString()])
  return {
    canEditExpenses: state.getIn(['application', 'user', 'canEditExpenses']),
    token: state.getIn(['application', 'token']),
    expenseId: expense ? expense.get('id') : null,
    description: expense ? expense.get('description') : '',
    amount: expense ? expense.get('amount') : 0,
    paid_at: expense ? expense.get('paid_at') : '',
  }
}

const mapDispatchToProps = (dispatch, ownProps) => {
  const expenseId = parseInt(ownProps.match.params.id, 10)
  const token = ownProps.match.params.token

  return {
    onClickDelete: (event, expenseId) => {
      event.preventDefault();
      dispatch(deleteExpense(token, expenseId));
    },
    onUpdateDescription: (description) => {
      dispatch(updateExpense(token, expenseId, {description}))
    },
    onUpdateAmount: (amount) => {
      dispatch(updateExpense(token, expenseId, {amount}))
    },
    onUpdatePaidAt: (paid_at) => {
      dispatch(updateExpense(token, expenseId, {paid_at}))
    },
  }
}

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseDetails)
