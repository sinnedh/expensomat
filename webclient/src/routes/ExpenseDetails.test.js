import React from 'react'
import { Map } from 'immutable'
import configureStore from 'redux-mock-store'
import ExpenseDetails from './ExpenseDetails'

describe('ExpenseDetails', () => {
  it('renders properly', () => {
    const initialState = Map({
      application: Map({ token: 'ABCD1234' }),
      expenses: Map({
        1: Map({id: 1, description: 'Expense 1', amount: 999, paid_at: '2017-11-11'}),
        2: Map({id: 2, description: 'Expense 2', amount: 1599, paid_at: '2016-12-08'}),
      })
    })
    const mockStore = configureStore()
    const store = mockStore(initialState)
    const wrapper = shallow(
      <ExpenseDetails
        match={{params: {id: 1, token: 'ABCD1234'}}}
        store={store}
        />
    )
    expect(wrapper.dive()).toMatchSnapshot()
  })
})
