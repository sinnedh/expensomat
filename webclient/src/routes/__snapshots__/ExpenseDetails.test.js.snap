// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExpenseDetails renders properly 1`] = `
<div>
  <h1>
    Expense details:
  </h1>
  <div>
    <label>
      Description:
    </label>
     
    <EditableInput
      isEditable={true}
      onClickSave={[Function]}
      value="Expense 1"
    />
  </div>
  <div>
    <label>
      Amount:
    </label>
     
    <EditableInput
      isEditable={true}
      onClickSave={[Function]}
      value={999}
    />
  </div>
  <div>
    <label>
      Paid at:
    </label>
     
    <EditableInput
      isEditable={true}
      onClickSave={[Function]}
      value="2017-11-11"
    />
  </div>
</div>
`;
